File: 1-100/98-validate-binary-search-tree/98-validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root, long long l = -1e10, long long r = 1e10) {
        if (root) {
            if (root->val <= l || root->val >= r) return false;
            return isValidBST(root->left, l, root->val) & isValidBST(root->right, root->val, r);
        }
        return true;
    }
};

File: 1-100/98-validate-binary-search-tree/98-validate-binary-search-tree.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public boolean isValidBST(TreeNode root) {
		if(root==null || (root.left==null && root.right==null)) return true;
		else return isValidBST(root, Long.MIN_VALUE, Long.MAX_VALUE);
    }
	
	private boolean isValidBST(TreeNode root, long min, long max) {
	    if(root==null) return true;
	    
	    if(!(min<root.val && root.val<max)) return false;
	    return isValidBST(root.left, min, root.val) && isValidBST(root.right, root.val, max);
	}
}